src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
+            info['entries'] = entries
+
+        self.send_update({'info': info}, reset=reset)
+
+    def process_pending_updates(self) -> None:
+        assert self.pending is not None
+        if self.pending:
+            self.process_update(self.pending)
+        self.pending = None
+
+    def interesting(self, name: str) -> bool:
+        if name == '':
+            return True
+        else:
+            # only report updates on entry filenames if we match them
+            return fnmatch.fnmatch(name, self.effective_fnmatch)
+
+    def schedule_update(self, name: str) -> None:
+        if not self.interesting(name):
+            return
+
+        if self.pending is None:
+            asyncio.get_running_loop().call_later(0.1, self.process_pending_updates)
+            self.pending = set()
+
+        self.pending.add(name)
+
+    def report_error(self, err: int) -> None:
+        if err == errno.ENOENT:
+            problem = 'not-found'
+        elif err in (errno.EPERM, errno.EACCES):
+            problem = 'access-denied'
+        elif err == errno.ENOTDIR:
+            problem = 'not-directory'
+        else:
+            problem = 'internal-error'
+
+        self.send_update({'error': {
+            'problem': problem, 'message': os.strerror(err), 'errno': errno.errorcode[err]
+        }}, reset=True)
+
+    def flag_onlydir_error(self, fd: Handle) -> bool:
+        # If our requested path ended with '/' then make sure we got a
+        # directory, or else it's an error.  open() will have already flagged
+        # that for us, but systemd_ctypes doesn't do that (yet).
+        if not self.watch or not self.path.endswith('/'):
+            return False
+
+        buf = os.fstat(fd)  # this should never fail
+        if stat.S_IFMT(buf.st_mode) != stat.S_IFDIR:
+            self.report_error(errno.ENOTDIR)
+            return True
+
+        return False
+
+    def report_initial_state(self, fd: Handle) -> None:
+        if self.flag_onlydir_error(fd):
+            return
+
+        self.fd = fd
+
+        entries = {''}
+        if self.fnmatch:
+            try:
+                entries.update(os.listdir(f'/proc/self/fd/{self.fd}'))
+                self.effective_fnmatch = self.fnmatch
+            except OSError:
+                # If we failed to get an initial list, then report nothing from now on
+                self.effective_fnmatch = ''
+
+        self.process_update({e for e in entries if self.interesting(e)}, reset=True)
+
+    def do_inotify_event(self, mask: InotifyEvent, cookie: int, rawname: 'bytes | None') -> None:
+        logger.debug('do_notify_event(%r, %r, %r)', mask, cookie, rawname)
+        name = (rawname or b'').decode(errors='surrogateescape')
+
+        self.schedule_update(name)
+
+        if name and mask | (InotifyEvent.CREATE | InotifyEvent.DELETE |
+                            InotifyEvent.MOVED_TO | InotifyEvent.MOVED_FROM):
+            # These events change the mtime of the directory
```

Comment:
Perhaps `and 'mtime' in attrs` (or a cached flag) as an optimization? As you were concerned about minimizing the events sent over the wire.

Reply:
This would have no impact on what's sent over the wire: if mtime isn't in the attrs, we'll schedule the update, do the `fstat()` on the directory, notice that nothing has changed, and go back to sleep.

We could save ourselves the wake-up and the stat, I suppose... but we also just changed a file in the directory, so we're going to be waking up to stat the file, anyway.

The only place this would make a difference is if we were monitoring a directory, without monitoring its mtime, and without fnmatch set, and entries were changing a lot in the directory.  That seems like a fairly slim edge case.

Oh: and it's not just `mtime`.  The `tag` attribute also changes if the timestamp is modified.  Which is enough to convince me that adding a special check here would be far too brittle.  I can imagine we add some other mtime-dependent attribute in the future and forget to come back here to update it.